### PR TITLE
Develop/gpt2

### PR DESCRIPTION
**Pull request by ifromeast**
This is the project to pre-train GPT2 on IPU.

### CODE REVIEW

File: pytorch/gpt2/README.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+This directory contains an implementation of GPT2 models in PyTorch for the IPU, leveraging the HuggingFace Transformers library. 
+
+There is one example for GPT2 pre-training: `train_gpt2.py`
+
+## Environment setup
+
+First, install the Poplar SDK following the instructions in the Getting Started guide for your IPU system. Make sure to source the `enable.sh` scripts for Poplar and PopART.
+
+SDK version: 2.4.0-EA.1+808_poptorch
```
Comment (philb-graphcore) [inline pytorch/gpt2/README.md:11]: We shouldn't be putting EA SDK numbers into the public domain.
If we are aligning this for the 2.4 release we should just say "SDK 2.4".

File: pytorch/gpt2/README.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    --enable-half-partials True \
+    --train-path <chosen-folder-for-preprocessed-files>/wikicorpus_en_one_article_per_line.pkl \
+    --epochs 3 \
+    --save-model-path './checkpoints/gpt2_medium'
+```
+## Run the pre-training application of GPT2-medium by poprun on POD128
+```console
+bash run/poprun_pretraining_POD128.sh
+```
+The model has not converge to SOTA, and we are still working on this.
```
Comment (philb-graphcore) [inline pytorch/gpt2/README.md:150]:
```suggestion
This model is under active development, and details on performance and convergence will be added soon. ```

File: pytorch/gpt2/README.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+The newest results will be updated once we got it.
```
Comment (philb-graphcore) [inline pytorch/gpt2/README.md:151]:
```suggestion
```

File: pytorch/gpt2/run/pretraining.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+time=$(date "+%Y%m%d%H%M%S")
+
+#export POPLAR_ENGINE_OPTIONS='{"profiler.includeFlopEstimates": "true","autoReport.all":"true","debug.allowOutOfMemory": "true", "autoReport.executionProfileProgramRunCount":"1", "autoReport.directory":"./logs/small_mlp_bs16"}'
```
Comment (philb-graphcore) [inline pytorch/gpt2/run/pretraining.sh:4]: Do we want to leave the profiling enabled for this?

**Review by philb-graphcore — CHANGES_REQUESTED** [review]
Looking good. A few minor things on the README & IP side I'd suggest we tidy up.

**Review by internetoftim — APPROVED** [review]
GPT2 M/L license review done

**Comment by internetoftim** [comment]
GPT2 M/L update for SDK 2.5

**Pull request merged**